[WIP] 2594: Add Google Calendar link

dorner opened the pull request:
<!--Read comments, before committing pull request read checklist again

# Checklist:

- I have performed a self-review of my own code,
- I have commented my code, particularly in hard-to-understand areas,
- I have made corresponding changes to the documentation,
- I have added tests that prove my fix is effective or that my feature works,
- New and existing unit tests pass locally with my changes ("bundle exec rake"),
- Title include "WIP" if work is in progress.

-->

Resolves #2594 

### Description
This is attempt number 2 at this feature (Attempt number 1 - direct sync - can still work, it just means a bit more work and more code to manage long-term). This will create a route that serves up an ICS calendar which can be added to Google Calendar (or any other calendar app that supports ICS).

The following changes had to be done to enable this:

* Add a time_zone attribute to storage locations. This defaults to America/LosAngeles. Right now the drop-down includes all time zones, alphabetized. I was thinking of limiting it to just North American time zones for more ease of use.
* Create the route that renders the ICS calendar. This route will show **all** distribution pickups - I was thinking of limiting it to those within the past year or in the future, to avoid having to pull every distribution since the dawn of time.
* The route needs to be publicly viewable for e.g. Google Calendar to see it. This means that we can't guarantee that a user is logged in at the time. However, I implemented a hash of the organization ID so that people can't just guess the URL, nor calculate it. I think this is probably enough security for this feature.

### Type of change

* New feature (non-breaking change which adds functionality)

### How Has This Been Tested?

Nothing yet, working on it :)

### Screenshots

<img width="1244" alt="image" src="https://user-images.githubusercontent.com/1986893/157966468-9531e604-2daa-496e-b0e9-0527267dd0eb.png">

<img width="1275" alt="image" src="https://user-images.githubusercontent.com/1986893/157965965-575efc18-9647-4dff-92ec-3360c52470bf.png">

dorner commented on app/services/calendar_service.rb:
```diff
@@ -0,0 +1,33 @@
+        e.dtend = Icalendar::Values::DateTime.new(dist.issued_at + 15.minutes, "tzid" => tz_id)
+        e.summary = "Pickup from #{dist.partner.name}"
+        e.location = dist.storage_location.address
+        e.url = "https://www.human-essentials.com/diaper_bank/distributions/schedule"
```
Not 100% sure this is the actual URL of the app?

dorner commented on db/migrate/20220311191446_add_time_zone_to_storage_locations.rb:
```diff
@@ -0,0 +1,7 @@
+class AddTimeZoneToStorageLocations < ActiveRecord::Migration[6.1]
+  def change
+    safety_assured do
+      add_column :storage_locations, :time_zone, :string, null: false, default: 'America/Los_Angeles'
```
According to strong_migrations, this is no longer unsafe if our Postgres version is >= 11 (which I assume it is).

scooter-dangle commented on app/views/storage_locations/_form.html.erb:
```diff
@@ -26,6 +26,9 @@
                 <%= f.input :warehouse_type, label: "Warehouse Type", wrapper: :input_group, collection: StorageLocation::WAREHOUSE_TYPES %>
+
+                <%= f.input :time_zone, label: "Time Zone", wrapper: :input_group,
+                            collection: ActiveSupport::TimeZone.all.sort_by(&:name).to_h { |z| [z.name, z.tzinfo.name]} %>
```
Note: Not necessarily advocating a specific change yet, but I think sorting all the timezones by name in whatever dropdown we use might be unwieldy for users. E.g., currently they'd see a dropdown like this:
![image](https://user-images.githubusercontent.com/934707/158030319-1506e6ff-b37b-46a4-983c-dbe105c9d67e.png)

Especially with the fact that most of the `z.name` values are cities, I was initially looking for `New York`, since that's what Linux (and macOS?) have me select when I manually select the time zone.

Possible improvements:
1) Having the identifier be UTC offset + name would make it easier to navigate east-to-west, and would co-locate most of the US+Canada zones that all(?) of our users would be looking for.
2) Instead of or in addition to the above, putting the US+Canada zones (including Hawaii + Alaska, _plus_ the special case zones (the non-NE portion of Arizona)) explicitly at the top of the list since all of our current users are in those.

dorner commented:
I've changed the time zone dropdown to show the American time zones first, west to east, and then all the rest.

<img width="1270" alt="image" src="https://user-images.githubusercontent.com/1986893/159082521-ce4e4614-2a0c-464b-b691-527a973fbfe1.png">

Updated the distributions to only show 1 past year plus future pickups.

I think this is good to review!

seanmarcia commented:
I'm going to merge and test on staging -- google still hasn't approved the google auth in prod yet.

seanmarcia commented:
Ah! Cool :) We'll have to add some instructions for that -- a lot of our folks are older so the easiest we can make it the better :)

dorner commented:
Yep, can do that in another PR. I wanted to test it and make sure it actually worked first - unfortunately this isn't something I could test on my local since I need a URL. :) Now that I say that I probably could have uploaded it to S3 with a public URL and tested it that way 🤔
